Extend scriptability patch 2

amosonn opened the pull request:
As suggested on https://github.com/OmniSharp/omnisharp-vim/pull/680 , I applied the same change to other actions which navigate: instead of always navigating, then calling the callback with the number of locations, if the callback is present navigation is skipped and the callback receives the list of locations.

The one place where I didn't change this is in definition: there the logic is split between ~~`locations.vim`~~ `definitions.vim` and `metadata.vim`, and callback-calling can occur from both. Since `metadata#Find` isn't actually used anywhere else, what to me seems the best course of action is:

- Merge `metadata.vim` into ~~`locations.vim`~~ `definitions.vim` as local functions
- Refactor this metadata logic "higher" in the call-graph, so that we have a point where we always get a location - either of a "real" file or of a temp file created from metadata
- Pass this location to either navigation or callback
- Remove callbacks completely from `Preview` (like in `implementations.vim`)

But I wanted to ask you before attempting that, maybe I missed something.

amosonn committed:
When callback present, don't navigate.

Callback receives full location, not just count.
Everywhere except for GotoDefinition.

amosonn committed:
Remove unused Callback

nickspoons commented on autoload/OmniSharp/actions/usings.vim:
```diff
@@ -4,14 +4,18 @@ set cpoptions&vim
   if g:OmniSharp_server_stdio
-    let Callback = function('s:CBFixUsings', [opts])
-    call s:StdioFix(Callback)
+    call s:StdioFind(Callback)
```
This should be `s/s:StdioFix`, not `s:StdioFind`

amosonn replied:
Done.

nickspoons commented:
This is great, nice to simplify these function signatures a bit!

I think that you're right that `metadata.vim` can be moved to local functions elsewhere, although I would think it makes sense to move them to `definition.vim`, where they are currently used, not `locations.vim`, is that what you meant?

I also agree about removing the callback from `OmniSharp#actions#definition#Preview`

I don't totally see what you mean with your 2nd point though, about refactoring the metadata handling higher in the call graph. Perhaps try making a commit so we can see how it looks? We can always decide not to exclude it if it is problematic.

amosonn commented:
Yeah, that is what I meant, it was already late :). I'll draft this change later and push it here.

amosonn committed:
Merge metadata.vim into definition.vim

amosonn committed:
Separate metadata fetching and navigation

amosonn committed:
Remove indirection of s:MetadataFind

amosonn committed:
Clean callback in definition

amosonn commented:
Added the change in multiple commits, to make it easier to follow. Fixed the mistake you found, and also updated the documentation in the first commit.

nickspoons commented:
Great stuff, thanks again @amosonn!
